src/main/kotlin/org/rust/ide/completion/DeriveCompletionProvider.kt

```diff
@@ -0,0 +1,63 @@
+package org.rust.ide.completion
+
+import com.intellij.codeInsight.completion.CompletionParameters
+import com.intellij.codeInsight.completion.CompletionProvider
+import com.intellij.codeInsight.completion.CompletionResultSet
+import com.intellij.codeInsight.lookup.LookupElementBuilder
+import com.intellij.patterns.ElementPattern
+import com.intellij.patterns.PlatformPatterns.psiElement
+import com.intellij.psi.PsiElement
+import com.intellij.util.ProcessingContext
+import org.rust.lang.RustLanguage
+import org.rust.lang.core.parser.RustPsiTreeUtil
+import org.rust.lang.core.psi.RustCompositeElementTypes
+import org.rust.lang.core.psi.RustMetaItemElement
+import org.rust.lang.core.psi.RustTokenElementTypes
+import org.rust.lang.core.resolve.util.RustResolveUtil
+
+class DeriveCompletionProvider : CompletionProvider<CompletionParameters>() {
```

Comment:
If I do this, then I can't declare `companion object ElementPatternFactory`.  This is my first time writing Kotlin, so I don't know if using a companion object here is appropriate anyway, but it seemed to work well for what I was doing.